gcc/rust/resolve/rust-toplevel-name-resolver-2.0.cc

```diff
@@ -71,6 +71,68 @@ TopLevel::visit (AST::Module &module)
 	      module.get_name ());
 }
 
+void
+TopLevel::visit (AST::ExternCrate &crate)
+{
+  CrateNum num;
+  if (!Analysis::Mappings::get ()->lookup_crate_name (
+	crate.get_referenced_crate (), num))
+    rust_unreachable ();
+
+  auto attribute_macros
+    = Analysis::Mappings::get ()->lookup_attribute_proc_macros (num);
+
+  auto bang_macros = Analysis::Mappings::get ()->lookup_bang_proc_macros (num);
+
+  auto derive_macros
+    = Analysis::Mappings::get ()->lookup_derive_proc_macros (num);
+
+  auto sub_visitor = [&] () {
+    if (derive_macros.has_value ())
+      for (auto &derive : derive_macros.value ())
+	{
+	  auto res = ctx.macros.insert (derive.get_trait_name (),
+					derive.get_node_id ());
+	  if (!res)
+	    {
+	      rust_error_at (UNKNOWN_LOCATION, ErrorCode::E0428,
+			     "macro %qs defined multiple times",
+			     derive.get_trait_name ().c_str ());
+	    }
+	}
+    if (attribute_macros.has_value ())
+      for (auto &attribute : attribute_macros.value ())
+	{
+	  auto res = ctx.macros.insert (attribute.get_name (),
+					attribute.get_node_id ());
+	  if (!res)
+	    {
+	      rust_error_at (UNKNOWN_LOCATION, ErrorCode::E0428,
+			     "macro %qs defined multiple times",
+			     attribute.get_name ().c_str ());
+	    }
+	}
+    if (bang_macros.has_value ())
+      for (auto &bang : bang_macros.value ())
+	{
+	  auto res = ctx.macros.insert (bang.get_name (), bang.get_node_id ());
+	  if (!res)
+	    {
+	      rust_error_at (UNKNOWN_LOCATION, ErrorCode::E0428,
+			     "macro %qs defined multiple times",
+			     bang.get_name ().c_str ());
+	    }
+	}
```

Comment:
I think we can factor this a little better since all the error messages are the same. Maybe a generic function which you call on `derive_macros`, `attribute_macros` and `bang_macros`?

Reply:
I'm not sure what you mean here. Should I create a function specifically to emit those errors ?

Reply:
Fixed in c085260734e.